File: front/src/lib/schemas/ml_projects.ts
"use client";

import { z } from "zod";

import { DatasetSchema } from "./datasets";
import { FoundationModelRunSchema } from "./foundation_models";
import { ModelRunSchema } from "./model_runs";
import { TagSchema } from "./tags";

// MLProject status enum - matches backend MLProjectStatus
export const MLProjectStatusSchema = z.enum([
  "draft",
  "active",
  "training",
  "inference",
  "completed",
  "archived",
]);

export type MLProjectStatus = z.infer<typeof MLProjectStatusSchema>;

// Main MLProject schema - matches backend MLProject schema
export const MLProjectSchema = z.object({
  uuid: z.string().uuid(),
  name: z.string(),
  description: z.string().nullable(),
  dataset_id: z.number().int().nullable().optional(),
  dataset: DatasetSchema.nullable().optional(),
  embedding_model_run_id: z.number().int().nullable().optional(),
  embedding_model_run: ModelRunSchema.nullable().optional(),
  foundation_model_id: z.number().int().nullable().optional(),
  status: MLProjectStatusSchema,
  default_similarity_threshold: z.number(),
  target_tags: z.array(TagSchema).default([]),
  created_by_id: z.string().uuid(),
  created_on: z.coerce.date(),
  project_id: z.string().optional(),
  // Stats
  dataset_scope_count: z.number().int().default(0),
  reference_sound_count: z.number().int().default(0),
  search_session_count: z.number().int().default(0),
  custom_model_count: z.number().int().default(0),
  inference_batch_count: z.number().int().default(0),
});

export type MLProject = z.infer<typeof MLProjectSchema>;

// Create schema - simplified, datasets are added in the Datasets tab
export const MLProjectCreateSchema = z.object({
  name: z.string().min(1, "Name is required"),
  description: z.string().optional(),
});

export type MLProjectCreate = z.infer<typeof MLProjectCreateSchema>;

// Update schema
export const MLProjectUpdateSchema = z.object({
  name: z.string().min(1).optional(),
  description: z.string().nullable().optional(),
  status: MLProjectStatusSchema.optional(),
  embedding_model_run_id: z.number().int().nullable().optional(),
  default_similarity_threshold: z.number().min(0).max(1).optional(),
});

export type MLProjectUpdate = z.infer<typeof MLProjectUpdateSchema>;

// Filter schema
export const MLProjectFilterSchema = z.object({
  dataset_id: z.number().int().positive().optional(),
  status: MLProjectStatusSchema.optional(),
});

// ============================================================================
// MLProject Dataset Scope
// ============================================================================

export const MLProjectDatasetScopeCreateSchema = z.object({
  dataset_uuid: z.string().uuid(),
  foundation_model_run_uuid: z.string().uuid(),
});

export const MLProjectDatasetScopeSchema = z.object({
  uuid: z.string().uuid(),
  dataset: DatasetSchema,
  foundation_model_run: FoundationModelRunSchema,
  created_on: z.coerce.date(),
});

File: old/back/src/echoroo/schemas/ml_projects.py
"""Schemas for ML Projects."""

import datetime
from enum import Enum
from uuid import UUID

from pydantic import BaseModel, Field

from echoroo.schemas.base import BaseSchema
from echoroo.schemas.datasets import Dataset
from echoroo.schemas.foundation_models import FoundationModel, FoundationModelRun
from echoroo.schemas.model_runs import ModelRun
from echoroo.schemas.tags import Tag

__all__ = [
    "MLProjectStatus",
    "MLProject",
    "MLProjectCreate",
    "MLProjectUpdate",
    "MLProjectStats",
    "MLProjectDatasetScope",
    "MLProjectDatasetScopeCreate",
]


class MLProjectStatus(str, Enum):
    """Status of an ML project - matches model definition."""

    SETUP = "setup"
    """Initial setup phase - defining targets and references."""

    SEARCHING = "searching"
    """Similarity search in progress."""

    LABELING = "labeling"
    """Reviewing and labeling search results."""

    TRAINING = "training"
    """Training a custom classifier."""

    INFERENCE = "inference"
    """Running inference on unlabeled data."""

    REVIEW = "review"
    """Reviewing model predictions."""

    COMPLETED = "completed"
    """Workflow completed, ready for export."""

    ARCHIVED = "archived"
    """Project archived and no longer active."""


class MLProjectCreate(BaseModel):
    """Schema for creating an ML project.

    Datasets are added separately via the Datasets tab after project creation.
    """

    name: str = Field(..., min_length=1, max_length=255)
    """Name of the ML project."""

    description: str | None = Field(default=None)
    """A description of the ML project goals and methodology."""


class MLProject(BaseSchema):
    """Schema for an ML project returned to the user."""

    uuid: UUID
    """UUID of the ML project."""

    id: int = Field(..., exclude=True)
    """Database ID of the ML project."""

    name: str
    """Name of the ML project."""

    description: str | None = None
    """A description of the ML project goals and methodology."""

    status: MLProjectStatus = MLProjectStatus.SETUP
    """Current status of the ML project."""

    project_id: str
    """Project identifier for access control."""

    foundation_model_id: int | None = None
    """Foundation model identifier used for embeddings."""

    foundation_model: FoundationModel | None = None
    """Hydrated foundation model information."""

    default_similarity_threshold: float = 0.8
    """Default threshold for similarity searches."""

    created_by_id: UUID
    """User who created the project."""

    target_tags: list[Tag] = Field(default_factory=list)
    """Tags that this project is targeting for detection."""

    dataset_scope_count: int = 0
    """Number of dataset scopes in the project."""

    reference_sound_count: int = 0
    """Number of reference sounds in the project."""

    search_session_count: int = 0
    """Number of search sessions conducted."""

    custom_model_count: int = 0
    """Number of custom models trained."""

    inference_batch_count: int = 0
    """Number of inference batches processed."""


class MLProjectUpdate(BaseModel):
    """Schema for updating an ML project."""

    name: str | None = Field(default=None, min_length=1, max_length=255)
    """Name of the ML project."""

    description: str | None = None
    """A description of the ML project goals and methodology."""

    status: MLProjectStatus | None = None
    """Current status of the ML project."""

    default_similarity_threshold: float | None = Field(
        default=None,
        ge=0.0,
        le=1.0,
    )
    """Default threshold for similarity searches."""


class MLProjectStats(BaseModel):
    """Aggregate statistics for an ML project."""

    total_reference_sounds: int = 0
    """Total number of reference sounds."""

    reference_sounds_by_tag: dict[str, int] = Field(default_factory=dict)
    """Count of reference sounds grouped by tag."""

    total_search_sessions: int = 0
    """Total number of search sessions."""

    total_labeled_results: int = 0
    """Total number of labeled search results."""

    positive_labels: int = 0
    """Number of results labeled as positive."""

    negative_labels: int = 0
    """Number of results labeled as negative."""

    total_custom_models: int = 0
    """Total number of custom models."""

    best_model_f1: float | None = None
    """Best F1 score achieved by custom models."""

    total_inference_batches: int = 0
    """Total number of inference batches."""

    total_predictions: int = 0
    """Total number of predictions made."""

    last_activity: datetime.datetime | None = None
    """Timestamp of the last activity in the project."""


class MLProjectDatasetScopeCreate(BaseModel):
    """Schema for adding a dataset scope to an ML project."""

    dataset_uuid: UUID = Field(..., description="UUID of the dataset to add")
    """UUID of the dataset to include in the project."""

    foundation_model_run_uuid: UUID = Field(
        ...,
        description="UUID of the foundation model run providing embeddings",
    )
    """UUID of the foundation model run that provides embeddings for this dataset."""


class MLProjectDatasetScope(BaseSchema):
    """Schema for an ML project dataset scope returned to the user."""

    uuid: UUID
    """UUID of the dataset scope."""

    dataset: Dataset
    """The dataset included in this scope."""

    foundation_model_run: FoundationModelRun
    """The foundation model run providing embeddings."""
